src/rgw/rgw_notify.cc

```diff
@@ -625,30 +675,76 @@ void populate_record_from_request(const req_state *s,
   set_event_id(record.id, etag, ts);
   record.bucket_id = s->bucket->get_bucket_id();
   // pass meta data
-  record.x_meta_map = s->info.x_meta_map;
+  if (s->info.x_meta_map.empty()) {
+    // try to fetch the metadata from the attributes
+    metadata_from_attributes(s, obj, record.x_meta_map);
+  } else {
+    record.x_meta_map = s->info.x_meta_map;
+  }
   // pass tags
-  record.tags = s->tagset.get_tags();
+  if (s->tagset.get_tags().empty()) {
+    // try to fetch the tags from the attributes
+    tags_from_attributes(s, obj, record.tags);
+  } else {
+    record.tags = s->tagset.get_tags();
+  }
   // opaque data will be filled from topic configuration
 }
 
-bool match(const rgw_pubsub_topic_filter& filter, const req_state* s, const rgw::sal::RGWObject* obj, EventType event) {
+bool match(const rgw_pubsub_topic_filter& filter, const req_state* s, rgw::sal::RGWObject* obj, 
```

Comment:
"match" is a pretty generic and context free name for what this function does?

Reply:
it is namespaced under `rgw::notify`.

the different "match()" functions have different prototypes according to the filter they use.
note that I used the term "match()" and not "filter()" so there is less ambiguity in whether you "filter out" or "filter in".

Reply:
the namespace isn't very visible here

Reply:
how about changing to `notification_match()` ?

then in `publish_reserve()` may be clearer:
```cpp
if (!notification_match(topic_filter, res.s, res.object, event_type, req_tags)) {
  // notification does not apply to req_state
  continue;
}
```